icu4j/main/classes/core/src/com/ibm/icu/text/PluralRules.java

```diff
@@ -1248,25 +1195,31 @@ static FixedDecimalSamples parse(String source) {
                 String[] rangeParts = TILDE_SEPARATED.split(range);
                 switch (rangeParts.length) {
                 case 1:
-                    FixedDecimal sample = new FixedDecimal(rangeParts[0]);
+                    DecimalQuantity sample =
+                        DecimalQuantity_DualStorageBCD.fromExponentString(rangeParts[0]);
                     checkDecimal(sampleType2, sample);
-                    samples2.add(new FixedDecimalRange(sample, sample));
+                    samples2.add(new DecimalQuantitySamplesRange(sample, sample));
                     break;
                 case 2:
-                    FixedDecimal start = new FixedDecimal(rangeParts[0]);
-                    FixedDecimal end = new FixedDecimal(rangeParts[1]);
+                    DecimalQuantity start =
+                            DecimalQuantity_DualStorageBCD.fromExponentString(rangeParts[0]);
+                    DecimalQuantity end =
+                            DecimalQuantity_DualStorageBCD.fromExponentString(rangeParts[1]);
                     checkDecimal(sampleType2, start);
                     checkDecimal(sampleType2, end);
-                    samples2.add(new FixedDecimalRange(start, end));
+                    samples2.add(new DecimalQuantitySamplesRange(start, end));
                     break;
                 default: throw new IllegalArgumentException("Ill-formed number range: " + range);
                 }
             }
-            return new FixedDecimalSamples(sampleType2, Collections.unmodifiableSet(samples2), bounded2);
+            return new DecimalQuantitySamples(sampleType2, Collections.unmodifiableSet(samples2), bounded2);
         }
 
-        private static void checkDecimal(SampleType sampleType2, FixedDecimal sample) {
-            if ((sampleType2 == SampleType.INTEGER) != (sample.getVisibleDecimalDigitCount() == 0)) {
+        private static void checkDecimal(SampleType sampleType2, DecimalQuantity sample) {
+            if ((sampleType2 == SampleType.INTEGER && sample.getPluralOperand(Operand.v) != 0) ||
+                    (sampleType2 == SampleType.DECIMAL && sample.getPluralOperand(Operand.v) == 0
+                            && sample.getPluralOperand(Operand.e) == 0
+                            && sample.getPluralOperand(Operand.n) != 0)) {  // is this double == 0 comparison fine? I think it is, just want to make sure.
```

Comment:
Question: You seem to be changing the truthiness of the expression. Why?

I can't answer the question about "is double == 0 fine" without understanding what this code is trying to accomplish that it couldn't accomplish before with the simpler expression.

Reply:
Done.